susemanager-utils/susemanager-sls/src/grains/cpe_grain.py

```diff
@@ -0,0 +1,78 @@
+import re
+
+def get_cpe_grain(grains):
+    ret = {"cpe": ""}
+
+    os_release = _parse_os_release("/etc/os-release")
+
+    cpe = os_release.get('CPE_NAME')
+    if cpe:
+        ret["cpe"] = cpe
+    else:
+        derived_cpe = _derive_cpe(grains)
+        if derived_cpe:
+            ret["cpe"] = derived_cpe
+
+    return ret
+
+# Copy-pasted from https://github.com/saltstack/salt/blame/master/salt/grains/core.py
+def _parse_os_release(*os_release_files):
+    """
+    Parse os-release and return a parameter dictionary
+
+    This function will behave identical to
+    platform.freedesktop_os_release() from Python >= 3.10, if
+    called with ("/etc/os-release", "/usr/lib/os-release").
+
+    See http://www.freedesktop.org/software/systemd/man/os-release.html
+    for specification of the file format.
+    """
+    # These fields are mandatory fields with well-known defaults
+    # in practice all Linux distributions override NAME, ID, and PRETTY_NAME.
+    ret = {"NAME": "Linux", "ID": "linux", "PRETTY_NAME": "Linux"}
+
+    errno = None
+    for filename in os_release_files:
+        try:
+            with open(filename, 'r') as ifile:
+                regex = re.compile("^([\\w]+)=(?:'|\")?(.*?)(?:'|\")?$")
+                for line in ifile:
+                    match = regex.match(line.strip())
+                    if match:
+                        # Shell special characters ("$", quotes, backslash,
+                        # backtick) are escaped with backslashes
+                        ret[match.group(1)] = re.sub(
+                            r'\\([$"\'\\`])', r"\1", match.group(2)
+                        )
+            break
+        except OSError as error:
+            errno = error.errno
+    else:
+        raise OSError(
+            errno, "Unable to read files {}".format(", ".join(os_release_files))
+        )
+
+    return ret
+
+
+def _derive_cpe(grains):
+    """
+    Try to derive the CPE of the system based on the collected core grains.
+
+    PS: This function is not guaranteed to derive the correct CPE as there could be a many 
+    variances of the same OS that require different CPEs, for example, release and beta versions.
```

Comment:
Do you have example os-release files that show the differences you're accounting for? Those would be good to test this code.

Reply:
This is an example of [RedHat 8 CPE](https://github.com/chef/os_release/blob/main/redhat_8), notice the 'GA' at the end. And this is the standard list of vulnerable RedHat 8 CPEs found in OVAL files:
```xml
   <affected_cpe_list>
    <cpe>cpe:/a:redhat:enterprise_linux:8</cpe>
    <cpe>cpe:/a:redhat:enterprise_linux:8::appstream</cpe>
    <cpe>cpe:/a:redhat:enterprise_linux:8::crb</cpe>
    <cpe>cpe:/a:redhat:enterprise_linux:8::highavailability</cpe>
    <cpe>cpe:/a:redhat:enterprise_linux:8::nfv</cpe>
    <cpe>cpe:/a:redhat:enterprise_linux:8::realtime</cpe>
    <cpe>cpe:/a:redhat:enterprise_linux:8::resilientstorage</cpe>
    <cpe>cpe:/a:redhat:enterprise_linux:8::sap</cpe>
    <cpe>cpe:/a:redhat:enterprise_linux:8::sap_hana</cpe>
    <cpe>cpe:/a:redhat:enterprise_linux:8::supplementary</cpe>
    <cpe>cpe:/o:redhat:enterprise_linux:8</cpe>
    <cpe>cpe:/o:redhat:enterprise_linux:8::baseos</cpe>
   </affected_cpe_list>
```

And this is [Ubuntu 22.04 CPE](https://nvd.nist.gov/products/cpe/detail/8DE10C8A-477C-42FF-8F96-D8813B729D3A?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Ao%3Acanonical%3Aubuntu_linux%3A*%3A*%3A*%3A*%3Alts%3A*%3A*%3A*&status=FINAL%2CDEPRECATED) with an LTS at the end. 

The code actually doesn't account for these differences. Since OVAL files for Ubuntu and Debian don't contain CPEs we derive them in the backend. For systems to be matched successfully, the derived CPE in salt must match the derived CPE in the backend.
https://github.com/uyuni-project/uyuni/blob/513ec82697c689d2fc638eaf1f614bc494f85dbb/java/code/src/com/suse/oval/vulnerablepkgextractor/DebianVulnerablePackagesExtractor.java#L82-L90